In [ ]:
%pip install numpy pandas seaborn matplotlib plotly scikit-learn nltk wordcloud pyspellchecker beautifulsoup4 prettytable

### Loading of Data

In [2]:
import pandas as pd
df = pd.read_csv(r'C:\Users\tiled\OneDrive\Desktop\DataSet\IMDB-Dataset.csv',encoding='latin-1')

In [3]:
df.head(10)["Reviews"]

0    *Disclaimer: I only watched this movie as a co...
1    I am writing this in hopes that this gets put ...
2    Really, I could write a scathing review of thi...
3    If you saw the other previous spoof movies by ...
4    This movie I saw a day early for free and I st...
5    Honestly, what is wrong with you, Hollywood? N...
6    I was given a free ticket to this film; so I c...
7    OK, so "Disastrous" isn't an imaginative barb ...
8    Jason Friedberg and Aaron Seltzer, the way eve...
9    Honestly the worst movie ever made. Theatre fu...
Name: Reviews, dtype: object

### Customizing Stopwords for Enhanced Text Preprocessing

*Enhancing the default stopword list by adding domain-specific words and removing negations to improve text analysis accuracy.*

In [4]:
from nltk.corpus import stopwords
import nltk

stop_words = stopwords.words('english')
nltk.download('averaged_perceptron_tagger')

new_stopwords = [
"would", "shall", "could", "might", "movie", "movies", "film", "films",
"cinema", "director", "directors", "actor", "actors", "actress", "actresses",
"cast", "screenplay", "plot", "story", "character", "characters", "scene",
"scenes", "soundtrack", "soundtracks","sequel", "prequel", "adaptation", "trailer", "genre", "genres", "release",
"theaters", "theatre", "performance", "performances", "acting", "blockbuster",
"indie", "oscars", "oscar", "award-winning", "box", "office", "premiere",
"rating", "ratings", "reviews", "animation", "cinematography", "editing", "script", "narrative",
"direction", "score", "soundtrack"
]

stop_words.extend(new_stopwords)
stop_words.remove("not")

[nltk_data] Downloading package averaged_perceptron_tagger to
[nltk_data]     C:\Users\tiled\AppData\Roaming\nltk_data...
[nltk_data]   Package averaged_perceptron_tagger is already up-to-
[nltk_data]       date!


In [5]:
stop_words=set(stop_words)
print(len(stop_words))

233


### Data Cleaning and Text Preprocessing Pipeline

*Comprehensive functions for cleaning and preparing text data, including removing special characters, URLs, stopwords, and expanding contractions to enhance the quality of the dataset.*

### Correction of Typos

*Written text often contains errors, such as “Fen” instead of “Fan.” To rectify these errors, a dictionary is employed to map words to their correct forms based on similarity.*

In [6]:
from spellchecker import SpellChecker

spell = SpellChecker()

def correct_typos(content):
    """
    Parameters:
    - content (str): The input text to be corrected.

    Returns:
    - str: The text with corrected typos.
    """
    try:
        words = content.split()
        corrected_words = []
        misspelled = spell.unknown(words)

        for word in words:
            if word in misspelled:
                corrected_word = spell.correction(word)
                corrected_words.append(corrected_word)
            else:
                corrected_words.append(word)
        
        corrected_content = ' '.join(corrected_words)
        return corrected_content
    except Exception as e:
        print(f"Error in correcting typos: {e}")
        return content

### Mapping and Replacement

*This involves mapping words to standardized language equivalents. For instance, words like “b4” and “ttyl,” commonly understood by humans as “before” and “talk to you later,” pose challenges for machines. Normalization entails mapping such words to their standardized counterparts.*

In [7]:
import re

def load_mapping_dictionary():
    """
    Returns:
    - dict: A dictionary mapping non-standard words to standardized words.
    """
    mapping_dict = {
        "b4": "before",
        "ttyl": "talk to you later",
        "u": "you",
        "r": "are",
        "lol": "laughing out loud",
        "idk": "i do not know",
        "btw": "by the way",
        "omg": "oh my god",
        "imo": "in my opinion",
        "np": "no problem",
        "gr8": "great",
        "l8r": "later",
        "gtg": "got to go",
        "thx": "thanks",
        "pls": "please",
        "plz": "please",
        "bc": "because",
        "cuz": "because",
        "y'all": "you all",
        "luv": "love",
        "wanna": "want to",
        "gonna": "going to",
        "hafta": "have to",
        "kinda": "kind of",
        "sorta": "sort of",
        "gimme": "give me",
        "lemme": "let me",
        "whatcha": "what are you",
        "whaddaya": "what do you",
    }
    return mapping_dict

def normalize_text(content, mapping_dict):
    """
    Parameters:
    - content (str): The input text to be normalized.
    - mapping_dict (dict): A dictionary mapping non-standard words to standardized words.
    
    Returns:
    - str: The normalized text.
    """
    try:
        words = content.split()
        normalized_words = []
        
        for word in words:
            clean_word = re.sub(r'[^\w\s]', '', word.lower())
            if clean_word in mapping_dict:
                replacement = mapping_dict[clean_word]
                normalized_words.append(replacement)
            else:
                normalized_words.append(word)
        
        normalized_content = ' '.join(normalized_words)
        return normalized_content
    except Exception as e:
        print(f"Error in normalizing text: {e}")
        return content

### Expanding Contractions

*Convert contractions like “don’t” to “do not” or “I’ll” to “I will.”*

In [8]:
contractions = {
        r"won't": "will not",
        r"can't": "cannot",
        r"i'm": "i am",
        r"ain't": "is not",
        r"let's": "let us",
        r"ma'am": "madam",
        r"shan't": "shall not",
        r"n't": " not",
        r"'re": " are",
        r"'s": " is",
        r"'d": " would",
        r"'ll": " will",
        r"'ve": " have",
        r"'m": " am",
        r"he's": "he is",
        r"she's": "she is",
        r"it's": "it is",
        r"that's": "that is",
        r"there's": "there is",
        r"who's": "who is",
        r"what's": "what is",
        r"where's": "where is",
        r"when's": "when is",
        r"why's": "why is",
        r"how's": "how is",
        r"would've": "would have",
        r"could've": "could have",
        r"should've": "should have",
        r"might've": "might have",
        r"must've": "must have",
        r"wouldn't": "would not",
        r"couldn't": "could not",
        r"shouldn't": "should not",
        r"mightn't": "might not",
        r"mustn't": "must not",
        r"don't": "do not",
        r"doesn't": "does not",
        r"didn't": "did not",
        r"hasn't": "has not",
        r"haven't": "have not",
        r"hadn't": "had not",
        r"can't've": "cannot have",
        r"shan't've": "shall not have",
        r"wouldn't've": "would not have",
        r"couldn't've": "could not have",
        r"shouldn't've": "should not have",
        r"mightn't've": "might not have",
        r"mustn't've": "must not have",
        r"i'd": "i would",
        r"i'll": "i will",
        r"i've": "i have",
        r"i'm": "i am",
        r"you'd": "you would",
        r"you'll": "you will",
        r"you've": "you have",
        r"you're": "you are",
        r"he'd": "he would",
        r"he'll": "he will",
        r"he's": "he is",
        r"she'd": "she would",
        r"she'll": "she will",
        r"she's": "she is",
        r"it'd": "it would",
        r"it'll": "it will",
        r"it's": "it is",
        r"they'd": "they would",
        r"they'll": "they will",
        r"they're": "they are",
        r"we'd": "we would",
        r"we'll": "we will",
        r"we're": "we are",
        r"there'd": "there would",
        r"that'd": "that would",
        r"who'd": "who would",
        r"who'll": "who will",
        r"who're": "who are",
        r"what've": "what have",
        r"where've": "where have",
        r"when've": "when have",
        r"why've": "why have",
        r"how've": "how have",
        r"lets": "let us",
        r"lets'": "let us",
        r"gonna": "going to",
        r"wanna": "want to",
        r"gotta": "got to",
        r"oughtn't": "ought not",
        r"needn't": "need not",
        r"daren't": "dare not",
        r"maam": "madam",
        r"gimme": "give me",
        r"lemme": "let me",
        r"whatcha": "what are you",
        r"whaddaya": "what do you",
        r"gimme": "give me",
        r"gonna": "going to",
        r"gotta": "got to",
        r"hafta": "have to",
        r"wanna": "want to",
        r"ain't": "am not",
        r"y'all": "you all",
        r"coulda": "could have",
        r"woulda": "would have",
        r"shoulda": "should have",
        r"'bout": "about",
        r"'til": "until",
        r"can't've": "cannot have",
        r"could've": "could have",
        r"might've": "might have",
        r"must've": "must have",
        r"should've": "should have",
        r"would've": "would have",
        r"you'd've": "you would have",
    }

def contraction_expansion(content):
    """
    Parameters:
    - content (str): The input text.

    Returns:
    - str: Text with expanded contractions.
    """
    
    for contraction, expanded in contractions.items():
        pattern = re.compile(contraction, flags=re.IGNORECASE)
        content = pattern.sub(expanded, content)
    return content

### Removing Accents and Diacritics
*Sometimes, people use accented characters like é, ö, etc. to signify emphasis on a particular letter during pronunciation. Normalize text by removing accents and diacritical marks from characters.*

In [9]:
import unicodedata

def remove_accents(content):
    """
    Parameters:
    - content (str): The input text to be normalized.
    
    Returns:
    - str: The text with accents and diacritics removed.
    """
    try:
        normalized = unicodedata.normalize('NFKD', content)
        without_accents = ''.join([c for c in normalized if not unicodedata.combining(c)])
        return without_accents
    except Exception as e:
        print(f"Error in removing accents: {e}")
        return content

### Removing Extra Whitespace

*Normalize text by removing extra spaces and leading/trailing spaces*

In [10]:
def remove_extra_whitespace(text):
    """
    Parameters:
    - text (str): The input text to be normalized.
    
    Returns:
    - str: The text with extra spaces removed.
    """
    return ' '.join(text.split())

### Eliminating HTML Tags

*In cases where raw text originates from sources such as web scraping or screen capture, it often carries along HTML tags. These tags introduce unwanted noise and contribute little to the comprehension and analysis of the text. Therefore, it becomes necessary to strip them.*

In [11]:
from bs4 import BeautifulSoup

def remove_html_tags(text):
    """
    Parameters:
    - text (str): The input text containing HTML tags.
    
    Returns:
    - str: The text with HTML tags removed.
    """
    soup = BeautifulSoup(text, "html.parser")
    return soup.get_text()

### Handling URLs

*Frequently, individuals include URLs, particularly in social media content, to supplement context with additional information. However, URLs tend to vary across samples and can be considered noise.*

In [12]:
import re 

def remove_url(text):
    return re.sub(r"(https|http)?:\S*", "", text)

### Case Standardising and Removing Special Characters

*Special characters are non-alphanumeric characters. The characters like %,$,&, etc are special. In most NLP tasks, these characters add no value to text understanding and induce noise into algorithms. We can use regular expressions to remove special characters.*

In [13]:
def remove_special_character(content):
    content = content.lower()
    return re.sub(r'\W+',' ', content)

### Removing Stopwords

*In most cases, stopwords like I, am, me, etc. don’t add any information that can help in modeling. Keeping them in the text introduces unnecessary noise and can significantly increase the dimensionality of feature vectors, which can negatively impact both computation cost and model accuracy.*

In [14]:
def remove_stopwords(content):
    clean_data = []
    for i in content.split():
        if i.strip().lower() not in stop_words and i.strip().lower().isalpha():
            clean_data.append(i.strip().lower())
    return " ".join(clean_data)

In [15]:
def data_cleaning(content):
    # content = correct_typos(content)
    mapping_dict = load_mapping_dictionary()
    content = remove_extra_whitespace(content)
    content = normalize_text(content, mapping_dict)
    content = contraction_expansion(content)
    content = remove_url(content)
    content = remove_accents(content)
    # content = remove_html_tags(content)
    content = remove_special_character(content)
    content = remove_stopwords(content)
    return content

### Applying Data Cleaning to Reviews and Displaying Results

In [16]:
%%time
pd.options.display.max_colwidth = 1000
df['Reviews_clean']=df['Reviews'].apply(data_cleaning)
df.head(5)

CPU times: total: 53.3 s
Wall time: 3min 14s


Ratings  \
0      1.0   
1      1.0   
2      1.0   
3      1.0   
4      1.0   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   Reviews  \
0  *Disclaimer: I only watched this movie as a conditional agreement. And I see films for free. I wouldn't be caught dead giving my hard earned money to these idiots.Well, to explain the depth of this 'film', I could write my shortest review, ever. Don't see this movie. It is by far the stupidest, lamest, most lazy, and unbelievably UNFUNNY movie I have ever seen. It is a total disaster. But since my hatred for this movie, and the others like it, extends far beyond one viewing, I think I'll go on for a bit.I don't know any of the people in the movie besides Carmen Electra, Vanessa Minnillo, and Kim Kardashian, but it doesn't matter. They're all horrible, though I think that was the point. The editing is flat out horrible, and possibly blatant continuity errors make this crapfast even crappier than I thought it would be. Now I know that these films are not supposed to be serious at all, but come on, it's film-making 101 that if someone gets a minor facial cut, it should be there in the...   
1  I am writing this in hopes that this gets put over the previous review of this "film". How anyone can find this slop entertaining is completely beyond me. First of all a spoof film entitled "Disaster Movie", should indeed be a spoof on disaster films. Now I have seen 1 (yes count them, 1) disaster film being spoofed, that being "Twister". How does Juno, Iron Man, Batman, The Hulk, Alvin and the Chipmunks, Amy Winehouse, or Hancock register as Disaster films? Selzterwater and Failburg once again have shown that they lack any sort of writing skill and humor. Having unfortunately been tortured with Date Movie and Epic Movie I know exactly what to expect from these two...no plot, no jokes just bad references and cheaply remade scenes from other films. Someone should have informed them that satire is more than just copy and paste from one film to another, though I shouldn't say that because some of these actually just seem to be taken from trailers.There is nothing clever or witty or re...   
2  Really, I could write a scathing review of this turd sandwich, but instead, I'm just going to be making a few observations and points I've deduced.There's just no point in watching these movies anymore. Does any reader out there remember Scary Movie? Remember how it was original with a few comedic elements to it? There was slapstick, some funny lines, it was a pretty forgettable comedy, but it was worth the price of admission. Well, That was the last time this premise was funny. STOP MAKING THESE MOVIES. PLEASE.I could call for a boycott of these pieces of monkey sh*t, but we all know there's going to be a line up of pre pubescent annoying little buggers, spouting crappy one liners like, "THIS IS SPARTA!" and, "IM RICK JAMES BITCH" so these movies will continue to make some form of monetary gain, considering the production value of this movie looks like it cost about 10 cents to make.

## Feature Engineering

In [17]:
df['Label'] = df['Ratings'].apply(lambda x: '1' if x >= 7 else ('0' if x<=4 else '2'))

df=df[df.Label<'2']
data=df[['Reviews_clean','Label']]

data['Label'].value_counts()

Label
0    60000
1    60000
Name: count, dtype: int64

### Lemmatization

In [18]:
from nltk.stem import WordNetLemmatizer
from nltk import word_tokenize  

nltk.download('wordnet')
nltk.download('punkt')

class LemmaTokenizer(object):
    def __init__(self):
        self.wordnetlemma = WordNetLemmatizer()
    
    def __call__(self, reviews):
        """
        Lemmatizes a list of tokens to their base form.

        Parameters:
        - tokens (list): A list of words to be lemmatized.

        Returns:
        - list: A list of lemmatized words.
        """
        return [self.wordnetlemma.lemmatize(word) for word in word_tokenize(reviews)]

[nltk_data] Downloading package wordnet to
[nltk_data]     C:\Users\tiled\AppData\Roaming\nltk_data...
[nltk_data]   Package wordnet is already up-to-date!
[nltk_data] Downloading package punkt to
[nltk_data]     C:\Users\tiled\AppData\Roaming\nltk_data...
[nltk_data]   Package punkt is already up-to-date!


## Text Vectorization with CountVectorizer and TF-IDF

In [25]:
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import TfidfVectorizer

In [20]:
train_data, test_data = train_test_split(data, test_size=0.3, random_state=42, shuffle=True)

In [21]:
y_train = train_data['Label']
y_test = test_data['Label']

In [22]:
import warnings
warnings.filterwarnings('ignore')

In [26]:
def vectorize_text(train_data, test_data, tokenizer, ngram_range=(1, 1), min_df=10, max_features=500):
    """
    Vectorizes text data using specified vectorizer type.

    Parameters:
    - train_data (pd.Series): The training text data to be vectorized.
    - test_data (pd.Series): The test text data to be vectorized.
    - tokenizer (callable): A tokenizer function or class instance.
    - ngram_range (tuple): The range of n-grams to consider.
    - min_df (int): Minimum document frequency for terms.
    - max_features (int): Maximum number of features to consider.

    Returns:
    - np.ndarray: The vectorized training text data.
    - np.ndarray: The vectorized test text data.
    - Vectorizer: The fitted vectorizer instance (TfidfVectorizer).
    """
  
    vectorizer = TfidfVectorizer(analyzer="word", 
                                 tokenizer=tokenizer, 
                                 ngram_range=ngram_range, 
                                 min_df=min_df,
                                 max_features=max_features)
  

    x_train_vectorized = vectorizer.fit_transform(train_data).toarray()
    x_test_vectorized = vectorizer.transform(test_data).toarray()

    return x_train_vectorized, x_test_vectorized, vectorizer

### Testing Logistic Regression with TF-IDF Vectorizers: N-gram Range 1 to 3

**Vectorization with TfidfVectorizer (N-gram Range: 1 to 3)**

In [28]:
x_train_ngrams_tfidf_vectorized, x_test_ngrams_tfidf_vectorized, ngram_tfidf_vectorizer = vectorize_text(
    train_data['Reviews_clean'], 
    test_data['Reviews_clean'], 
    tokenizer=LemmaTokenizer(), 
    ngram_range=(1, 3), 
    min_df=10, 
    max_features=5000
)

## Model Selection 

In [38]:
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.ensemble import AdaBoostClassifier
from sklearn.pipeline import Pipeline

### Logistic Regression Model

In [30]:
model_1=LogisticRegression()

In [31]:
%%time
model_1.fit(x_train_ngrams_tfidf_vectorized,y_train)

CPU times: total: 2min 48s
Wall time: 23.8 s


LogisticRegression()

**Evaluation on Test and Train dataset**

In [32]:
from sklearn.metrics import f1_score, roc_auc_score, precision_score, recall_score

***Performance Metrics***

1. **Precision Score:**
   - **Definition:** Precision is the ratio of correctly predicted positive observations to the total predicted positives. It answers the question: "Of all the instances that were predicted as positive, how many were actually positive?"
   - **Formula:** 
     - `Precision = True Positives / (True Positives + False Positives)`
   - **Interpretation:** A high precision score indicates a low false positive rate.
   

2. **AUC Score (Area Under the ROC Curve):**
   - **Definition:** AUC measures the ability of the model to distinguish between classes. The ROC curve is a plot of the true positive rate against the false positive rate at various threshold settings.
   - **Range:** 0 to 1, where 1 indicates a perfect model and 0.5 suggests no discriminative power.
   - **Interpretation:** A higher AUC indicates better model performance.

3. **F1 Score:**
   - **Definition:** The F1 score is the harmonic mean of precision and recall. It considers both false positives and false negatives.
   - **Formula:** 
     - `F1 Score = 2 * (Precision * Recall) / (Precision + Recall)`
   - **Interpretation:** A high F1 score indicates a balance between precision and recall.
   - **Note on Recall:** Recall (or Sensitivity) is the ratio of correctly predicted positive observations to all actual positives. It answers the question: "Of all the instances that were actually positive, how many were predicted as positive?"

***Training Dataset Metrics***

In [35]:
precision_score_training = precision_score(y_train, model_1.predict(x_train_ngrams_tfidf_vectorized), average='micro')
recall_score_training = recall_score(y_train, model_1.predict(x_train_ngrams_tfidf_vectorized), average='micro')
auc_score_training = roc_auc_score(y_train, model_1.predict_proba(x_train_ngrams_tfidf_vectorized)[:, 1], multi_class='ovo', average='macro')
f1_score_training = f1_score(y_train, model_1.predict(x_train_ngrams_tfidf_vectorized), average="weighted")

print("\n**Training Dataset:**")
print(f"- **Precision Score**: {precision_score_training}")
print(f"- **Recall Score**: {recall_score_training}")
print(f"- **AUC Score**: {auc_score_training}")
print(f"- **F1 Score**: {f1_score_training}")


**Training Dataset:**
- **Precision Score**: 0.90925
- **Recall Score**: 0.90925
- **AUC Score**: 0.9681792289615996
- **F1 Score**: 0.9092487265586905


***Test Dataset Metrics***

In [36]:
precision_score_test = precision_score(y_test, model_1.predict(x_test_ngrams_tfidf_vectorized), average='micro')
recall_score_test = recall_score(y_test, model_1.predict(x_test_ngrams_tfidf_vectorized), average='micro')
auc_score_test = roc_auc_score(y_test, model_1.predict_proba(x_test_ngrams_tfidf_vectorized)[:, 1], multi_class='ovo', average='macro')
f1_score_test = f1_score(y_test, model_1.predict(x_test_ngrams_tfidf_vectorized), average="weighted")

print("\n**Test Dataset:**")
print(f"- **Precision Score**: {precision_score_test}")
print(f"- **Recall Score**: {recall_score_test}")
print(f"- **AUC Score**: {auc_score_test}")
print(f"- **F1 Score**: {f1_score_test}")


**Test Dataset:**
- **Precision Score**: 0.8913333333333333
- **Recall Score**: 0.8913333333333333
- **AUC Score**: 0.9575016853183461
- **F1 Score**: 0.8913354584211626


### Decision Tree Classifier

In [39]:
model_2 = Pipeline(
    steps=[
    ("classifier", DecisionTreeClassifier())
    ]
)

In [40]:
%%time
model_2.fit(x_train_ngrams_tfidf_vectorized, y_train)

Pipeline(steps=[('classifier', DecisionTreeClassifier())])

***Training Dataset Metrics for Decision Tree Classifier***

In [41]:
precision_score_training_dt = precision_score(y_train, model_2.predict(x_train_ngrams_tfidf_vectorized), average='micro')
recall_score_training_dt = recall_score(y_train, model_2.predict(x_train_ngrams_tfidf_vectorized), average='micro')
auc_score_training_dt = roc_auc_score(y_train, model_2.predict_proba(x_train_ngrams_tfidf_vectorized)[:, 1], multi_class='ovo', average='macro')
f1_score_training_dt = f1_score(y_train, model_2.predict(x_train_ngrams_tfidf_vectorized), average="weighted")

print("### Decision Tree Classifier Performance Metrics")
print("\n**Training Dataset:**")
print(f"- **Precision Score**: {precision_score_training_dt}")
print(f"- **Recall Score**: {recall_score_training_dt}")
print(f"- **AUC Score**: {auc_score_training_dt}")
print(f"- **F1 Score**: {f1_score_training_dt}")

### Decision Tree Classifier Performance Metrics

**Training Dataset:**
- **Precision Score**: 0.9999404761904762
- **Recall Score**: 0.9999404761904762
- **AUC Score**: 0.9999999770404832
- **F1 Score**: 0.9999404762037628


***Test Dataset Metrics for Decision Tree Classifier***

In [42]:
precision_score_test_dt = precision_score(y_test, model_2.predict(x_test_ngrams_tfidf_vectorized), average='micro')
recall_score_test_dt = recall_score(y_test, model_2.predict(x_test_ngrams_tfidf_vectorized), average='micro')
auc_score_test_dt = roc_auc_score(y_test, model_2.predict_proba(x_test_ngrams_tfidf_vectorized)[:, 1], multi_class='ovo', average='macro')
f1_score_test_dt = f1_score(y_test, model_2.predict(x_test_ngrams_tfidf_vectorized), average="weighted")

print("\n**Test Dataset:**")
print(f"- **Precision Score**: {precision_score_test_dt}")
print(f"- **Recall Score**: {recall_score_test_dt}")
print(f"- **AUC Score**: {auc_score_test_dt}")
print(f"- **F1 Score**: {f1_score_test_dt}")


**Test Dataset:**
- **Precision Score**: 0.7158888888888889
- **Recall Score**: 0.7158888888888889
- **AUC Score**: 0.7160518300211375
- **F1 Score**: 0.7158939083592555


*Observation*

The Decision Tree Classifier demonstrates near-perfect performance on the training dataset, with precision, recall, and AUC scores all close to 1, indicating potential overfitting. However, the significant drop in these metrics on the test dataset suggests that the model struggles to generalize well to unseen data, with precision, recall, and AUC scores around 0.716, indicating moderate performance.

### Decision Tree Classifier with max depth 11 to fix overfit 

In [52]:
model_3 = Pipeline(
    steps=[
        ("classifier", DecisionTreeClassifier(criterion='gini', max_depth=11, min_samples_split=2, min_samples_leaf=1)),
    ]
)

In [49]:
%%time
model_3.fit(x_train_ngrams_tfidf_vectorized,y_train)

CPU times: total: 55.7 s
Wall time: 2min 10s


Pipeline(steps=[('classifier', DecisionTreeClassifier(max_depth=15))])

In [50]:
precision_score_training_dt_11 = precision_score(y_train, model_3.predict(x_train_ngrams_tfidf_vectorized), average='micro')
recall_score_training_dt_11 = recall_score(y_train, model_3.predict(x_train_ngrams_tfidf_vectorized), average='micro')
auc_score_training_dt_11 = roc_auc_score(y_train, model_3.predict_proba(x_train_ngrams_tfidf_vectorized)[:, 1], multi_class='ovo', average='macro')
f1_score_training_dt_11 = f1_score(y_train, model_3.predict(x_train_ngrams_tfidf_vectorized), average="weighted")

print("### Decision Tree Classifier with max depth 11 Performance Metrics")
print("\n**Training Dataset:**")
print(f"- **Precision Score**: {precision_score_training_dt_11}")
print(f"- **Recall Score**: {recall_score_training_dt_11}")
print(f"- **AUC Score**: {auc_score_training_dt_11}")
print(f"- **F1 Score**: {f1_score_training_dt_11}")

### Decision Tree Classifier with max depth 11 Performance Metrics

**Training Dataset:**
- **Precision Score**: 0.771452380952381
- **Recall Score**: 0.771452380952381
- **AUC Score**: 0.8611728585743647
- **F1 Score**: 0.768116525928192


In [51]:
precision_score_test_dt_11 = precision_score(y_test, model_3.predict(x_test_ngrams_tfidf_vectorized), average='micro')
recall_score_test_dt_11 = recall_score(y_test, model_3.predict(x_test_ngrams_tfidf_vectorized), average='micro')
auc_score_test_dt_11 = roc_auc_score(y_test, model_3.predict_proba(x_test_ngrams_tfidf_vectorized)[:, 1], multi_class='ovo', average='macro')
f1_score_test_dt_11 = f1_score(y_test, model_3.predict(x_test_ngrams_tfidf_vectorized), average="weighted")

print("\n**Test Dataset:**")
print(f"- **Precision Score**: {precision_score_test_dt_11}")
print(f"- **Recall Score**: {recall_score_test_dt_11}")
print(f"- **AUC Score**: {auc_score_test_dt_11}")
print(f"- **F1 Score**: {f1_score_test_dt_11}")


**Test Dataset:**
- **Precision Score**: 0.7209722222222222
- **Recall Score**: 0.7209722222222222
- **AUC Score**: 0.7797682749007329
- **F1 Score**: 0.7174349318364087
